Applications/Spire/Source/OrderImbalanceIndicator/CachedOrderImbalanceIndicatorModel.cpp

```diff
@@ -0,0 +1,108 @@
+#include "Spire/OrderImbalanceIndicator/CachedOrderImbalanceIndicatorModel.hpp"
+#include "Spire/Spire/QtPromise.hpp"
+
+using namespace boost::icl;
+using namespace boost::posix_time;
+using namespace Nexus;
+using namespace Spire;
+
+CachedOrderImbalanceIndicatorModel::CachedOrderImbalanceIndicatorModel(
+  std::shared_ptr<OrderImbalanceIndicatorModel> m_source_model)
+  : m_source_model(std::move(m_source_model)) {}
+
+OrderImbalanceIndicatorModel::SubscriptionResult
+    CachedOrderImbalanceIndicatorModel::subscribe(
+    const boost::posix_time::ptime& start,
+    const boost::posix_time::ptime& end,
+    const OrderImbalanceSignal::slot_type& slot) {
+  if(contains(m_ranges, continuous_interval<ptime>::closed(start, end))) {
+    return get_subscription(start, end, slot);
+  }
+  return load_imbalances(start, end, slot);
+}
+
+OrderImbalanceIndicatorModel::SubscriptionResult
+    CachedOrderImbalanceIndicatorModel::get_subscription(
+    const boost::posix_time::ptime& start,
+    const boost::posix_time::ptime& end,
+    const OrderImbalanceSignal::slot_type& slot) {
+  m_subscriptions.push_back(Subscription(start, end));
+  auto imbalances = std::vector<OrderImbalance>();
+  std::copy_if(m_imbalances.begin(), m_imbalances.end(),
+    std::back_inserter(imbalances), [=] (auto& imbalance) {
+      return start <= imbalance.m_timestamp && imbalance.m_timestamp <= end;
+    });
+  return {m_subscriptions.back().m_imbalance_signal.connect(slot),
+    QtPromise([requested_imbalances = std::move(imbalances)] {
+      return std::move(requested_imbalances);
+    })};
+}
+
+OrderImbalanceIndicatorModel::SubscriptionResult
+    CachedOrderImbalanceIndicatorModel::load_imbalances(
+    const boost::posix_time::ptime& start,
+    const boost::posix_time::ptime& end,
+    const OrderImbalanceSignal::slot_type& slot) {
+  auto promises = std::vector<QtPromise<std::vector<OrderImbalance>>>();
+  auto ranges = interval_set<ptime>(
+    {continuous_interval(start, end)}) - m_ranges;
+  for(auto& range : ranges) {
+    auto [connection, promise] = m_source_model->subscribe(range.lower(),
+      range.upper(), [=] (auto& imbalance) {
+        on_order_imbalance(imbalance);
+      });
+    m_connections.push_back(std::move(connection));
+    promises.push_back(std::move(promise));
+  }
+  m_subscriptions.push_back(Subscription(start, end));
+  return {m_subscriptions.back().m_imbalance_signal.connect(slot),
+    all(std::move(promises)).then([=] (auto& imbalances_lists) {
+      m_ranges.add({start, end});
+      auto loaded_imbalances_lists = imbalances_lists.Get();
```

Comment:
Line 61 makes a copy of a ```vector```, instead of making a copy iterate over the expression directly:

    for(auto& list : imbalances_lists.Get())